Enable redirection from landing page to record view page

tylerjmchugh opened the pull request:
Addresses issue #8949

Emails with links currently go to the "metadata api view" which is a restricted format which was intended to be used in browsers with limited capabilities. i.e. PDF printer. This is problematic for users as they want the links to go to the record view page where they can perform actions on the metadata.

This PR aims to fix this issue by updating the "metadata api view" to accept a `recordViewFormatter` parameter. When this parameter is present the "metadata api view" will redirect to the regular "record view page" and using the specified formatter. A configuration for the formatter was added to the ui configuration.

A configuration for the formatter used in email links was also added to the ui configuration:
<img width="800" height="131" alt="image" src="https://github.com/user-attachments/assets/cde30e26-821d-4876-a17e-a7aa3f12fac8" />

If the value is not empty the recordViewFormatter parameter will be added to the links in the emails.

Assuming the only defined formatter is `defaultView`:
- `/{portal}/api/records/{metadataUuid}` - Displays the "metadata api view"
- `/{portal}/api/records/{metadataUuid}?recordViewFormatter=defaultView` - Redirects to the "record view page" using the `defaultView` formatter
- `/{portal}/api/records/{metadataUuid}?recordViewFormatter=undefinedView` - Falls back to the "metadata api view"
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [X] I have read the [contribution guidelines](https://github.com/geonetwork/core-geonetwork/blob/main/CONTRIBUTING.md)
- [X] *Pull request* provided for `main` branch, backports managed with label
- [X] *Good housekeeping* of code, cleaning up comments, tests, and documentation
- [X] *Clean commit history* broken into understandable chucks, avoiding big commits with hundreds of files, cautious of reformatting and whitespace changes
- [X] *Clean commit message*s, longer verbose messages are encouraged
- [ ] *API Changes* are identified in commit messages
- [ ] *Testing* provided for features or enhancements using [automatic tests](https://github.com/geonetwork/core-geonetwork/blob/main/software_development/TESTING.md)
- [ ] *User documentation* provided for new features or enhancements in [manual](https://github.com/geonetwork/core-geonetwork/tree/main/docs/manual)
- [ ] *Build documentation* provided for development instructions in `README.md` files
- [ ] *Library management* using `pom.xml` dependency management. Update build documentation with intended library use and library tutorials or documentation

ianwallen commented on services/src/main/java/org/fao/geonet/api/records/MetadataApi.java:
```diff
@@ -179,6 +182,10 @@ public String getRecord(
         @PathVariable
         String metadataUuid,
+        @Parameter(description = "Formatter to use on the record view page. " +
+            "If not specified, no redirect to the record view page is performed.")
```
May also want to spec that invalid values are also ignored.
```suggestion
            "If invalid or not specified, no redirect to the record view page is performed.")
```

josegar74 commented:
@tylerjmchugh Is there any reason to add this as a user interface setting rather than an application setting? As a user interface setting, if different subportals are defined with customised user interface settings, it is necessary to configure it in each user interface. I am not sure if that is the intention.

ianwallen commented:
@josegar74 
It does seems to make sense to have it per ui so that each portal could have different settings.  If in the system configuration then it means that all portals needs to be configured with the same formatter to be used in the email.  If using `defaultView` then that is ok since they will most likely all have this but if creating a new formatter specifically for emails then it would need to be created in each ui.

I think it would be better if a checkbox was added to the formatter. i.e. 

<img width="838" height="411" alt="image" src="https://github.com/user-attachments/assets/b6db14df-68cd-483d-98e7-48cc853e4dd6" />

It would be better if this could be a radio buttons but that may be difficult to implement.  so maybe we just get first item checked.

What could be done is to add a option in the system configuration to identify the default formatter to be used in portal ui configurations.  Then whenever someone configures a new ui it could automatically check that formatter.  But again, this works well for `defaultView` and `full` which are part of the default ui configuration.  But if someone specified `email` as the formatter then that would do nothing since the `email` formatter does not exists in the ui template and it would need to be manually added.  So this option would only be good to formatters that are part of the ui template.

Rosspetcsiro commented on core/src/main/java/org/fao/geonet/util/LocalizedEmailComponent.java:
```diff
@@ -361,12 +362,18 @@ private String replaceLinks(String message) {
+        // Set the recordViewFormatter parameter if emailRecordViewFormatter is present
+        String emailRecordViewFormatterParameter = StringUtils.isNotBlank(emailRecordViewFormatter) ? "?recordViewFormatter=" + emailRecordViewFormatter : "";
+
         String newPlaceholder;
         if (replaceLinksWithHtmlFormat) {
             newPlaceholder = "{{index:uuid}}";
         } else {
             newPlaceholder = "'{{'index:uuid'}}'";
         }
-        return message.replace("{{link}}", settingManager.getNodeURL() + "api/records/" + newPlaceholder);
+        return message.replace("{{link}}", settingManager.getNodeURL() + "api/records/" + newPlaceholder + emailRecordViewFormatterParameter);
```
Is it possible to also add something like this for the metadata permalink genaration?

<img width="874" height="373" alt="Image" src="https://github.com/user-attachments/assets/ee9f4f7f-889d-4fb3-b9f5-189ca39906ce" />

<img width="1200" height="239" alt="Image" src="https://github.com/user-attachments/assets/0e6238e1-04b7-4b61-8a09-453bb37b50e2" />

So that it also does not point to the API. 

I think logic is here: https://github.com/geonetwork/core-geonetwork/blob/main/core/src/main/java/org/fao/geonet/kernel/datamanager/base/BaseMetadataUtils.java#L328-L348

tylerjmchugh commented:
@ianwallen @josegar74

In my latest push I have updated to use a radio button on the record view formatter config:
<img width="3084" height="1281" alt="image" src="https://github.com/user-attachments/assets/58714502-df91-4acf-90e2-29f3ca886a6d" />

---

Additionally I have updated the `gnUiConfigHelp` directive to use a `<div>` instead of a `<p>` so that `<pre><code>` snippets can be properly embedded:

Before:
<img width="450"  alt="image" src="https://github.com/user-attachments/assets/cc9cc1e5-4712-4da9-bf9f-45dd3f67d00a" />
After:
<img width="450" alt="image" src="https://github.com/user-attachments/assets/650ce95a-94d7-472c-b4f2-bec3cbfe5de5" />

--- 

And as mentioned in the [above comment](https://github.com/geonetwork/core-geonetwork/pull/8966#discussion_r2323171857) I have implemented the parameter for the default permalink.

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/datamanager/base/BaseMetadataUtils.java:
```diff
@@ -344,7 +345,25 @@ public String getPermalink(String uuid, String language) {
+
+            // If the recordLinkFormatter is defined append it to the defaultLink
+            if (StringUtils.isNotBlank(recordLinkFormatter)) {
+                link += ((defaultLink.contains("?") ? '&' : '?') + "recordViewFormatter=" + recordLinkFormatter);
```
Maybe can be replaced by:

```java
link = UriComponentsBuilder.fromUriString(defaultLink).queryParam("recordLinkFormatter", recordLinkFormatter).build().toString();
```

josegar74 reviewed (approved):
A small optional suggestion, other than that looks fine.
